OpenRA.Mods.Common/Traits/World/Locomotor.cs

```diff
@@ -190,80 +193,148 @@ public WorldMovementInfo GetWorldMovementInfo(World world)
 			return new WorldMovementInfo(world, this);
 		}
 
-		public int MovementCostToEnterCell(WorldMovementInfo worldMovementInfo, Actor self, CPos cell, Actor ignoreActor = null, CellConditions check = CellConditions.All)
-		{
-			var cost = MovementCostForCell(worldMovementInfo.World, worldMovementInfo.TerrainInfos, cell);
-			if (cost == int.MaxValue || !CanMoveFreelyInto(worldMovementInfo.World, self, cell, ignoreActor, check))
-				return int.MaxValue;
-			return cost;
-		}
+		public virtual bool DisableDomainPassabilityCheck { get { return false; } }
+
+		public virtual object Create(ActorInitializer init) { return new Locomotor(init.Self, this); }
+	}
 
-		public SubCell GetAvailableSubCell(
-			World world, Actor self, CPos cell, SubCell preferredSubCell = SubCell.Any, Actor ignoreActor = null, CellConditions check = CellConditions.All)
+	public class Locomotor : IWorldLoaded
+	{
+		struct CellCache
 		{
-			if (MovementCostForCell(world, cell) == int.MaxValue)
-				return SubCell.Invalid;
+			public readonly CellBlocking CellBlocking;
+			public readonly short Cost;
+			public readonly LongBitSet<PlayerBitMask> Blocking;
 
-			if (check.HasCellCondition(CellConditions.TransientActors))
+			public CellCache(short cost, LongBitSet<PlayerBitMask> blocking, CellBlocking cellBlocking)
 			{
-				Func<Actor, bool> checkTransient = otherActor => IsBlockedBy(self, otherActor, ignoreActor, check);
+				Cost = cost;
+				Blocking = blocking;
+				CellBlocking = cellBlocking;
+			}
 
-				if (!SharesCell)
-					return world.ActorMap.AnyActorsAt(cell, SubCell.FullCell, checkTransient) ? SubCell.Invalid : SubCell.FullCell;
+			public CellCache UpdateCost(short cost)
```

Comment:
Do you think `WithCost`/`WithBlocking` be more standard naming here - these are returning new structs, not updating the current one.